File: CODING/ARRAY/quick_sort.py
def partition(arr,piovt):
    pivot=arr[-1]
    i=-1
    j=0
    while(j<(len(arr)-1)):
        if(arr[j]<=pivot):
            i+=1
            arr[j],arr[i]=arr[i],arr[j]
        j+=1
    
    arr[i+1],arr[-1]=arr[-1],arr[i+1]
    return [arr,i+1]


def quicksort(arr,left,right):
    if left<right:
        return arr
    pivot=right
    arr,part=partition(arr,pivot)
    x=quicksort(arr,left,part-1)
    y=quicksort(arr,part+1,right)
    return x+y

# print(partition([4,2,6,9,8,7],7))
print(quicksort([4,2,6,9,8,7],0,6))

File: CODING/sorting_algos/quick_sort.cpp
#include<bits/stdc++.h>

using namespace std;

void printarr(int arr[],int size){
    for(int i=0;i<size;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}
int partition(int arr[],int start,int end){
    int i=start-1;
    int j=start;
    int pivot_index=end;
    while(j<end){
        if(arr[j]<arr[pivot_index]){
            i++;
            swap(arr[i],arr[j]);
        }
        j++;
    }
    swap(arr[i+1],arr[pivot_index]);
    return i+1;
}


void quick_sort_util(int arr[],int start,int end){
    
    if(start<end){
    
    
    int pivot=partition(arr,start,end);
    quick_sort_util(arr,start,pivot-1);
    quick_sort_util(arr,pivot+1,end);
    
    }
}
void quick_sort(int arr[],int size){
    quick_sort_util(arr,0,size-1);
}

int  main()
{
    int arr[]={5,1,2,9,3,8,7,6,2,1,7};
    int size=sizeof(arr)/sizeof(arr[0]);
    cout<<size<<endl;
    printarr(arr,size);
    quick_sort(arr,size);
    printarr(arr,size); 
}